### PR TITLE
Rest arguments

### LINKED ISSUE #7306: [Feature Request / Lua] Type for handling Lua var args "..."
I am using Haxe to make World Of Warcraft addons and something that is used frequently is the Lua var args as a parameter "..." my proposal is to add a specific Type for this

In Haxe I would write
```haxe
eventManager.RegisterEvent("MESSAGE", (self:Frame, event:String, args:VArgs) -> {
    trace(args.length);
    if (event == "MESSAGE") {
        var msg:String = args[0];
    }
});
```
and that would compile to something like this
```lua
eventManager:RegisterEvent("MESSAGE", function(self, event, ...)
    print(select("#", ...));
    if (event == "MESSAGE") then
        local msg = select(0, ...);
    end
end);
```

Limitations would be that you can only place it in a parameter list and only as the last parameter.

### PR DESCRIPTION
**Pull request by RealyUniqueName**
**Closed in favor of #9961**

Implements rest arguments as an abstract over targets native corresponding features.
Closes #7306
Softly deprecates `haxe.extern.Rest` (no warning for now).

- [x] js
- [x] python
- [x] php
- [ ] flash
- [ ] cpp
- [ ] eval
- [x] lua
- [ ] java
- [ ] cs
- [ ] jvm
- [ ] neko

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
@jdonaldson Could you please add a lua implementation?

File: std/haxe/Rest.hx [hunk 1]
```diff
@@ -0,0 +1,58 @@
+package haxe;
+
+import haxe.iterators.RestKeyValueIterator;
+import haxe.iterators.RestIterator;
+
+@:coreType
+abstract Rest<T> {
```
Comment (Gama11) [inline std/haxe/Rest.hx:4]: `to Iterable<T>` for static extension support might be nice?

Comment (RealyUniqueName) [reply std/haxe/Rest.hx:4]: No, because it's not compatible with `Iterable`.
The only thing we can to is adding `@:to Iterable`

Comment (Gama11) [reply std/haxe/Rest.hx:4]: Hm... but it has `iterator()`?

Comment (RealyUniqueName) [reply std/haxe/Rest.hx:4]: But it only exists at compile time, while `Iterable` needs `iterator` to exist at run time.

**Comment by jdonaldson** [comment]
Looks like a good start... I got things working for Lua in your test cases.  

I think we'll need to add some more restrictions for Rest.... here's some things that are possible, but add a lot of complications to support:

1. Returning a Rest type might need special behavior
2. Passing a Rest type as a dynamic argument.

These should probably be restricted, with a helpful compiler message:
1. Passing a Rest type as an argument before other standard arguments. 
2. Passing two or more Rest types

I'm not handling any of these cases yet

**Comment by RealyUniqueName** [comment]
> 1. Passing a Rest type as an argument before other standard arguments.
> 2. Passing two or more Rest types

These are handled by the compiler

File: std/lua/_std/haxe/Rest.hx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return select(index+1, this);
+    }
+
+	@:to public inline function toArray():Array<T> {
+		return [for(i in 0...length) get(i)];
+	}
+
+    public inline function toString() : String {
+        return this.toArray().toString();
+    }
```
Comment (Aurel300) [inline std/lua/_std/haxe/Rest.hx]: Inconsistent tabs vs spaces in this file (maybe others?)

Comment (Gama11) [reply std/lua/_std/haxe/Rest.hx]: It's probably time to run haxe-formatter over everything again. :)

**Comment by jdonaldson** [comment]
> > 1. Passing a Rest type as an argument before other standard arguments.
> > 2. Passing two or more Rest types
> 
> These are handled by the compiler

There's not a very clear/explicit error message for these cases... 

```haxe
function rest(r:Rest<Int>, o :Int){
  return o;
}
[...]
var r = rest(3, 2, 1, 0);
```

or

```haxe
function rest(r:Rest<Int>, o : Rest<Int>){
  trace(r + " is the value for r");
  return o;
}
```

Gives:

```
Main.hx|8 col 16 error| Int should be haxe.Rest<Int>
Main.hx|8 col 16 error| For function argument 'r'
```

File: std/lua/_std/haxe/Rest.hx (hunk 2 above)
Comment (jdonaldson) [reply std/lua/_std/haxe/Rest.hx]: I retabbed, sorry about that.

**Comment by RealyUniqueName** [comment]
> There's not a very clear/explicit error message for these cases...

Looks like it's only checked for fields. Will add for local functions too.

File: std/lua/_std/haxe/Rest.hx [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package haxe;
+import lua.Lua.select;
+import lua.TableTools.maxn;
+import lua.TableTools.pack;
+import lua.Table;
+
```
Comment (Aurel300) [inline std/lua/_std/haxe/Rest.hx:7]: Spaces -> Tabs again, to be consistent with the stdlib.

Comment (Gama11) [reply std/lua/_std/haxe/Rest.hx:7]: Or just run haxe-formatter over it, there's other formatting issues too. :D

Comment (jdonaldson) [reply std/lua/_std/haxe/Rest.hx:7]: FWIW I added haxe-formatter to this vim plugin : https://github.com/Chiel92/vim-autoformat/pull/297

I have it set to autorun when I open a haxe file now.

**Comment by jdonaldson** [comment]
Should we support returning a Rest argument?  I added a breaking test for it.

**Comment by nadako** [comment]
So... does this support calling rest-argument functions by passing `Array` (or something that can be created from one) as the rest argument? (i.e. `exec(a, b, myRestArgs)`). This would be required for such a feature because we have inheritance and `super` calls (also constructors).

**Comment by RealyUniqueName** [comment]
> So... does this support calling rest-argument functions by passing Array (or something that can be created from one) as the rest argument?

Not yet. I have to figure out inlining for this case first. Or disable inlining functions with rest args completely.

**Comment by Simn** [comment]
This PR will need an update.

Also, we have to stop randomly stopping to work on things like these.

**Comment by RealyUniqueName** [comment]
This was not planned and I've started this because I thought it wouldn't take a lot of effort, but it turned out I was wrong, so I moved back to planned tasks.

Main issue is an absence of syntax for passing a collection as rest arguments  (e.g. `callWithRest(...args)`)

**Comment by ALANVF** [comment]
Curious why Neko isn't supported so far. It supports variadic functions and dynamic function application, no?

**Comment by RealyUniqueName** [comment]
Closed in favor of #9961

**Pull request closed without merging**